alloydb/notebooks/batch_embeddings_update.ipynb

```diff
@@ -0,0 +1,1098 @@
+      },
+      "outputs": [],
+      "source": [
+        "# Max token count for the embeddings API\n",
+        "max_tokens = 20000\n",
+        "\n",
+        "# For some tokenizers and text, there's a rough approximation that 1 token corresponds to about 3-4 characters. This is a very general guideline and can vary significantly.\n",
+        "max_char_count = max_tokens * 3\n",
+        "\n",
+        "cols_to_embed = ['analysis','overview']\n",
+        "\n",
+        "# Model to use for generating embeddings\n",
+        "model_name = 'text-embedding-004'\n",
+        "\n",
+        "# Generate optimised embeddings for a given task\n",
+        "# Ref: https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/task-types#supported_task_types\n",
+        "task = \"SEMANTIC_SIMILARITY\""
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "D3FUBaXIUquR"
+      },
+      "source": [
+        "This runs the complete embeddings workflow:\n",
+        "\n",
+        "1. Gettting source data\n",
+        "2. Batching source data\n",
+        "3. Generating embeddings for batches\n",
+        "4. Updating data batches in the original table"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "syO1Zq3o5PnI",
+        "outputId": "8db5edfc-7b9e-46da-bda8-123444033b37"
+      },
+      "outputs": [],
+      "source": [
+        "import vertexai\n",
+        "import time\n",
+        "from vertexai.language_models import TextEmbeddingModel\n",
+        "\n",
+        "pool_size = 10\n",
+        "embed_data_concurrency = 20\n",
+        "batch_update_concurrency = 10\n",
+        "total_char_count = 0\n",
+        "\n",
+        "# Set up connections to the database\n",
+        "pool = await init_connection_pool(connector, database_name, pool_size=pool_size)\n",
+        "\n",
+        "# Initialise VertexAI and the model to be used to generate embeddings\n",
+        "vertexai.init(project=project_id, location=region)\n",
+        "model = TextEmbeddingModel.from_pretrained(model_name)\n",
+        "\n",
+        "start_time = time.monotonic()\n",
+        "\n",
+        "# Fetch source data from the database\n",
+        "source_data = get_source_data(pool, cols_to_embed)\n",
+        "\n",
+        "# Divide the source data into batches for efficient processing\n",
+        "batch_data = batch_source_data(source_data, cols_to_embed)\n",
+        "\n",
+        "# Generate embeddings for the batched data concurrently\n",
+        "embeddings_data = embed_objects_concurrently(\n",
+        "    cols_to_embed, batch_data, model, task, max_concurrency=embed_data_concurrency\n",
+        ")\n",
+        "\n",
+        "# Update the database with the generated embeddings concurrently\n",
+        "await batch_update_rows_concurrently(\n",
```

Comment:
As a user I would kind of expect just to run 1 method to generate the embeddings, compared to having to get and batch the source data then generate my embeddings, then updating the database. I have to copy around the variable cols_to_embed a lot. We might be able to simplify this devex more.

Reply:
Users can declare cols to embed (and other variables) and can now use the embeddings workflow while being able to configure it with params like pool_size, embed_concurrency and batch_update_concurrency. 

This would allow for a better devex.
Another alternative is to let each function use the global cols_to_embed variable and eliminate it's argument. Is that a better idea?

Reply:
+1 to keeping it as an argument